### PR TITLE
Step.Compile: add options struct for `addCSourceFiles`

### LINKED ISSUE #17410: change `addCSourceFiles` to be future proof by taking an options struct and add dependency option
```zig
pub const AddCSourceFilesOptions = struct {
    /// When provided, `files` are relative to `dependency` rather than the package that owns the `Compile` step.
    dependency: ?*std.Build.Dependency = null,
    files: []const []const u8,
    flags: []const []const u8 = &.{},
};

/// Handy when you have many C/C++ source files and want them all to have the same flags.
pub fn addCSourceFiles(self: *Compile, options: AddCSourceFilesOptions) void {
    // ...
}
```

Make it not repeat the C flags on the command line for every file even when a `dependency` is provided.

https://github.com/ziglang/zig/blob/e6590fea19e3eab94b35bfd3c36e29b53cefcaaf/lib/std/Build/Step/Compile.zig#L925

### PR DESCRIPTION
**Pull request by der-teufel-programming**
Closes #17410

### CODE REVIEW

File: lib/std/Build/Step/Compile.zig [hunk 1]
```diff
@@ -1549,8 +1559,14 @@ fn make(step: *Step, prog_node: *std.Progress.Node) !void {
                     try zig_args.append("--");
                     prev_has_cflags = true;
                 }
-                for (c_source_files.files) |file| {
-                    try zig_args.append(b.pathFromRoot(file));
+                if (c_source_files.dependency) |dep| {
+                    for (c_source_files.files) |file| {
+                        try zig_args.append(b.pathFromRoot(dep.path(file).getPath(b)));
```
Comment (Parzival-3141) [inline lib/std/Build/Step/Compile.zig:1564]: Shouldn't this be using `try zig_args.append(dep.builder.pathFromRoot(file));` instead?

**Comment by andrewrk** [comment]
Can you show the `--verbose-cc` output for a build.zig script that uses this API with several C files and several flags?

**Comment by der-teufel-programming** [comment]
https://gist.github.com/der-teufel-programming/546a8f228c7374dd2ac8f85c615f143a
This gist contains slightly formatted `--verbose` output for https://github.com/andrewrk/orca-ui-hello-zig with this small diff applied to the `build.zig`
```diff
-    for (libc_shim_files) |libc_shim_file| {
-        app.addCSourceFile(.{
-            .file = orca_dep.path(libc_shim_file),
-            .flags = cflags,
-        });
-    }
+    app.addCSourceFiles(.{
+        .dependency = orca_dep,
+        .files = libc_shim_files,
+        .flags = cflags,
+    });
```

**Comment by andrewrk** [comment]
This is not a good example because it has only 1 flag.

File: lib/std/Build/Step/Compile.zig (hunk 1 above)
Comment (andrewrk) [reply lib/std/Build/Step/Compile.zig:1564]: Yes, that's correct

Comment (der-teufel-programming) [reply lib/std/Build/Step/Compile.zig]: Done

**Comment by andrewrk** [comment]
Nice work, thank you!

**Comment by glingy** [comment]
I realize this is closed, but this is a breaking change that makes the build.zig file not compatible with 0.11.0, and also makes 0.12.0 not able to use old build.zig files which is going to potentially cause issues if using packages set up to build for 0.11.0 with a new zig project in the near future using 0.12.0. Can I request that the build.zig file be backwards compatible in the future so this isn't an issue and packages can slowly adopt the new syntax without breaking compatibility?

**Comment by glingy** [comment]
The reason being I just ran across a build.zig file that wasn't building on 0.12.0-dev, but was on my local computer (running 0.11.0), and discovered this change.

**Comment by andrewrk** [comment]
You can check the zig version in your build script and call the older API.

**Pull request merged**